### PR TITLE
Add gradual typing constraint definitions

### PR DESCRIPTION
**Pull request by migeed-z**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80147
* #80110
* #80102
* #80095
* #80084
* __->__ #79912

Differential Revision: [D37315020](https://our.internmc.facebook.com/intern/diff/D37315020)

Add high-level gradual type constraint definitions. Those definitions will be used by the constraint generator. The generator will generate high-level constraints for FX graphs which will then be transformed to constraints that are understood by Z3

### CODE REVIEW

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

File: torch/fx/experimental/migrate_gradual_types/constraint.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+class Conj(Constraint):
+    def __init__(self, conjuncts):
+        """
+        :param conjuncts: Conjuction of constraints
+        """
+        self.conjucts = conjuncts
+
+    def __eq__(self, other):
+        if isinstance(other, Conj):
+            return self.conjucts == other.conjucts and self.conjucts == other.conjucts
```
Comment (anijain2305) [inline torch/fx/experimental/migrate_gradual_types/constraint.py:19]: Just trying to learn - why do you have `self.conjucts == other.conjucts` twice?

**Review by anijain2305 — APPROVED** [review]

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

**Review by jamesr66a — APPROVED** [review]
Seems fine to me. Do you mind adding a description of what the PR is adding in the description field?

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37315020).

**Pull request closed without merging**